### PR TITLE
Vert.x TCCL fixes

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java [hunk 1]
```diff
@@ -253,9 +186,23 @@ public static Vertx initialize(VertxConfiguration conf, VertxOptionsCustomizer c
...
+        VertxThreadFactory vertxThreadFactory = new VertxThreadFactory() {
+            @Override
+            public VertxThread newVertxThread(Runnable target, String name, boolean worker, long maxExecTime,
+                    TimeUnit maxExecTimeUnit) {
+                var thread = VertxThreadFactory.INSTANCE.newVertxThread(target, name, worker, maxExecTime, maxExecTimeUnit);
+                thread.setContextClassLoader(tccl);
+                if (launchMode == LaunchMode.DEVELOPMENT) {
+                    devModeThreads.add(thread);
+                }
+                return VertxThreadFactory.super.newVertxThread(target, name, worker, maxExecTime, maxExecTimeUnit);
```
Comment (Sanne) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:199]: I suppose you meant to return the existing `thread` instance?

Comment (Sanne) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:199]: I mean, rather than invoking the factory once more.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java [hunk 2]
```diff
@@ -135,74 +132,9 @@ public void handle(AsyncResult<Void> event) {
...
-        final Object result;
-        try {
-            result = executorMethod.invoke(workerPool);
-        } catch (IllegalAccessException | InvocationTargetException e) {
-            throw new RuntimeException(e);
-        }
-        EnhancedQueueExecutor executor = (EnhancedQueueExecutor) result;
-        final Thread[] runningThreads = executor.getRunningThreads();
-        for (Thread t : runningThreads) {
+        for (var t : devModeThreads) {
```
Comment (Sanne) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:137]: Awesome :+1:

File: core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java [hunk 3]
```diff
@@ -383,7 +383,7 @@ private RunTimeConfigurationGenerator() {
             cc.getFieldCreator(C_BUILD_TIME_CONFIG_SOURCE)
                     .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL);
             final ResultHandle buildTimeConfigSource = clinit.newInstance(PCS_NEW, buildTimeValues,
-                    clinit.load("Build time config"), clinit.load(100));
+                    clinit.load("Build time config"), clinit.load(Integer.MAX_VALUE));
```
Comment (Sanne) [inline core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java:386]: How is this related?

File: core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java [hunk 4]
```diff
@@ -480,7 +480,7 @@ public void run() {
                             reinit.load(entry.getValue()));
                 }
                 final ResultHandle buildTimeConfigSource = reinit.newInstance(PCS_NEW, buildTimeValues,
-                        reinit.load("Build time config = Reloaded"), reinit.load(100));
+                        reinit.load("Build time config = Reloaded"), reinit.load(Integer.MAX_VALUE));
```
Comment (Sanne) [inline core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java:483]: How is this related?

Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java]: Oops, I had some unrelated changes locally and didn't notice them.

**Review by Sanne — APPROVED** [review]

**Comment by famod** [comment]
FTR, I confirm that this is fixing that nasty problem we were fighting with the last few days: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/main.3A.20NPE.20in.20SR.20Config.20ClassPathUtils/near/249911993

Thanks @stuartwdouglas, master of threads, contexts and loaders! 😁

**Comment by Sanne** [comment]
there's a potential race condition between expansion of the threadpool and the context being swapped: not a blocker for the current excellent PR, but we should follow up on that. 

Also considering some machines might want to start lots of threads, I'd replace the "copy on write set" with a regular set , as we get rid of the race with a lock.

**Comment by stuartwdouglas** [comment]
@Sanne race should be gone now, and it is also a lot more careful not to capture the original TCCL and cause a leak in dev mode.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java [hunk 5]
```diff
@@ -230,6 +176,10 @@ public void handle(AsyncResult<Void> event) {
             } catch (InterruptedException e) {
                 throw new RuntimeException(e);
             }
+            synchronized (devModeThreads) {
```
Comment (jaikiran) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:179]: Hello Stuart, is there a chance that the threads and classloader will leak if the previous try block throws an exception? In other words, should we use a `finally` block here to clear the collection and the classloader?

Comment (stuartwdouglas) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:179]: In theory this could happen, I will change it.

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java [hunk 6]
```diff
@@ -253,9 +204,34 @@ public static Vertx initialize(VertxConfiguration conf, VertxOptionsCustomizer c
 
         Vertx vertx;
 
+        Optional<ClassLoader> nonDevModeTccl;
```
Comment (famod) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:207]: This whole block looks very similar to `createThreadFactory()`, no?

IMO, the complexity warrants a single method that can be used in both cases, but YMMV!

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java [hunk 7]
```diff
@@ -509,18 +486,41 @@ public Integer get() {
         };
     }
 
-    public ThreadFactory createThreadFactory() {
-        ClassLoader tccl = Thread.currentThread().getContextClassLoader();
+    public ThreadFactory createThreadFactory(LaunchMode launchMode) {
+        Optional<ClassLoader> nonDevModeTccl;
+        if (launchMode == LaunchMode.DEVELOPMENT) {
+            currentDevModeNewThreadCreationClassLoader = Thread.currentThread().getContextClassLoader();
```
Comment (famod) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:492]: nitpicking: `currentDevModeNewThreadCreationClassLoader` says:
> This must be accessed under the {@link #devModeThreads} lock

Yet, I don't see a lock/syncronized here, probably because it's not needed because there is no concurrency here.

Comment (Sanne) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:492]: I was looking at the same :) And concluded that it's fine, but yea we could make it synchronized to make it more explicit.

Comment (famod) [reply extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:492]: ... or adjust the javadoc or leave an inline comment why it's not needed here (which brings us back to my other suggestion to extract a common method).

**Review by famod — APPROVED** [review]
LGTM and it still fixes the problem I was seeing. 🤓

**Comment by famod** [comment]
I briefly tested a few other things with my current app (e.g. dev mode reload) and it's all looking good so far.
